File: 2021/05/02/main.go
package main

import (
	"fmt"
	"sort"
)

type SortedList []int

func (list *SortedList) Insert(x int) {
	i := sort.SearchInts(*list, x)
	*list = append(*list, -1)
	copy((*list)[i+1:], (*list)[i:])
	(*list)[i] = x
}

func scheduleCourse(courses [][]int) int {
	qh := make(SortedList, 0)
	sort.Slice(courses, func(i, j int) bool { return courses[i][1] < courses[j][1] })
	start := 0
	for i := 0; i < len(courses); i++ {
		dura, end := courses[i][0], courses[i][1]
		start += dura
		qh.Insert(dura)
		if start > end {
			start -= qh[len(qh)-1]
			qh = qh[:len(qh)-1]
		}
	}
	return len(qh)
}

func main() {
	var courses = [][]int{{100, 200}, {200, 1300}, {1000, 1250}, {2000, 3200}}
	var r = scheduleCourse(courses)
	fmt.Println(r)
}

File: 2021/05/02/main.py
import heapq

class Solution:
    def scheduleCourse(self, courses: List[List[int]]) -> int:
        hp = []
        courses.sort(key=lambda x: x[1])
        cur = 0
        for a, b in courses:
            cur += a
            heapq.heappush(hp, -a)
            if cur > b:
                cur += heapq.heappop(hp)
        return len(hp)

if __name__ == '__main__':
    print('__main__')
    courses = [[3,2],[4,3]]#[[100,200],[200,1300],[1000,1250],[2000,3200]]
    res = Solution().scheduleCourse(courses)
    print(res)
